src/EditorFeatures/CSharp/MisplacedUsings/MisplacedUsingsCodeFixProvider.cs

```diff
@@ -0,0 +1,405 @@
+                        onBlankLine = false;
+                        break;
+
+                    case SyntaxKind.WhitespaceTrivia:
+                        fileHeaderBuilder.Add(firstTokenLeadingTrivia[i]);
+                        break;
+
+                    case SyntaxKind.EndOfLineTrivia:
+                        hasHeader = true;
+                        fileHeaderBuilder.Add(firstTokenLeadingTrivia[i]);
+
+                        if (onBlankLine)
+                        {
+                            done = true;
+                        }
+                        else
+                        {
+                            onBlankLine = true;
+                        }
+
+                        break;
+
+                    default:
+                        done = true;
+                        break;
+                }
+
+                if (done)
+                {
+                    break;
+                }
+            }
+
+            return hasHeader ? fileHeaderBuilder.ToImmutableArray() : ImmutableArray.Create<SyntaxTrivia>();
+        }
+
+        private static CompilationUnitSyntax AddFileHeader(CompilationUnitSyntax compilationUnit, ImmutableArray<SyntaxTrivia> fileHeader)
+        {
+            if (fileHeader.IsEmpty)
+            {
+                return compilationUnit;
+            }
+
+            // Add leading trivia to the first token.
+            var firstToken = compilationUnit.GetFirstToken(includeZeroWidth: true);
+            var newLeadingTrivia = firstToken.LeadingTrivia.InsertRange(0, fileHeader);
+            var newFirstToken = firstToken.WithLeadingTrivia(newLeadingTrivia);
+
+            return compilationUnit.ReplaceToken(firstToken, newFirstToken);
+        }
+
+        private class FixAll : DocumentBasedFixAllProvider
+        {
+            public static FixAllProvider Instance { get; } = new FixAll();
+
+            protected override string CodeActionTitle => CSharpEditorResources.Move_misplaced_using_directives;
+
+            /// <inheritdoc/>
+            protected override async Task<SyntaxNode> FixAllInDocumentAsync(FixAllContext fixAllContext, Document document, ImmutableArray<Diagnostic> diagnostics)
+            {
+                if (diagnostics.IsEmpty)
+                {
+                    return null;
+                }
+
+                var syntaxRoot = await document.GetSyntaxRootAsync(fixAllContext.CancellationToken).ConfigureAwait(false);
+                var options = await document.GetOptionsAsync(fixAllContext.CancellationToken).ConfigureAwait(false);
+                var newDocument = await GetTransformedDocumentAsync(document, (CompilationUnitSyntax)syntaxRoot, options, fixAllContext.CancellationToken).ConfigureAwait(false);
+                return await newDocument.GetSyntaxRootAsync(fixAllContext.CancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private class MoveMisplacedUsingsCodeAction : CodeAction
```

Comment:
we already have a CodeAction.DocumentChangeAction that we derive from for these cases.